lib/ClangImporter/ClangImporter.cpp

```diff
@@ -1024,6 +1024,46 @@ bool ClangImporter::isModuleImported(const clang::Module *M) {
   return M->NameVisibility == clang::Module::NameVisibilityKind::AllVisible;
 }
 
+bool ClangImporter::canImportModule(std::pair<Identifier, SourceLoc> moduleID) {
+  auto &clangContext = Impl.getClangASTContext();
+  auto &clangHeaderSearch = Impl.getClangPreprocessor().getHeaderSearchInfo();
+
+  // Look up the top-level module first, to see if it exists at all.
+  clang::Module *clangModule =
+    clangHeaderSearch.lookupModule(moduleID.first.str());
+  if (!clangModule)
+    return false;
+
+  if (isModuleImported(clangModule)) {
+    return true;
+  }
+
+  auto &rawDiagClient = Impl.Instance->getDiagnosticClient();
+  auto &diagClient = static_cast<ClangDiagnosticConsumer &>(rawDiagClient);
+
+  SmallVector<std::pair<clang::IdentifierInfo *, clang::SourceLocation>, 4>
+    clangPath;
+  clangPath.push_back({ &clangContext.Idents.get(moduleID.first.str()),
+                        Impl.exportSourceLoc(moduleID.second) });
+  auto importRAII = diagClient.handleImport(clangPath.front().first,
+                                            SourceLoc());
+
+  // FIXME: The source location here is completely bogus. It can't be
+  // invalid, and it can't be the same thing twice in a row, so we just use
+  // a counter. Having real source locations would be far, far better.
+  auto &srcMgr = clangContext.getSourceManager();
+  clang::SourceLocation clangImportLoc
+    = srcMgr.getLocForStartOfFile(srcMgr.getMainFileID())
+            .getLocWithOffset(Impl.ImportCounter++);
+  clang::ModuleLoadResult result =
+    Impl.Instance->loadModule(clangImportLoc, clangPath, clang::Module::Hidden,
+                              /*IsInclusionDirective=*/false);
+
+  // Now load the top-level module, so that we can check if the submodule
+  // exists without triggering a fatal error.
+  return (result != nullptr);
```

Comment:
`canImport` does not work with submodules at the moment.  I can make it work and see what happens.